### PR TITLE
fix: enable read gcs file from local

### PR DESCRIPTION
**Pull request by jesrypandawa**

### CODE REVIEW

File: dagger-functions/build.gradle [hunk 1]
```diff
@@ -64,7 +64,9 @@ dependencies {
     dependenciesFunctionsJar group: 'org.apache.commons', name: 'commons-jexl3', version: '3.1'
     dependenciesFunctionsJar group: 'org.isuper', name: 's2-geometry-library-java', version: '0.0.1'
     dependenciesFunctionsJar group: 'com.google.cloud', name: 'google-cloud-storage', version: '1.67.0'
-
+    dependenciesFunctionsJar group: 'org.apache.flink', name: 'flink-shaded-hadoop2-uber', version: '2.4.1-1.8.3'
+    dependenciesFunctionsJar group: 'com.google.cloud.bigdataoss', name: 'gcs-connector', version: '1.9.0-hadoop2'
```
Comment (MayurGubrele) [inline dagger-functions/build.gradle:68]: This jar already exists in /opt/flink/lib directory, just check once if we really need to add this

File: dagger-functions/build.gradle [hunk 2]
```diff
@@ -64,7 +64,9 @@ dependencies {
     dependenciesFunctionsJar group: 'org.apache.commons', name: 'commons-jexl3', version: '3.1'
     dependenciesFunctionsJar group: 'org.isuper', name: 's2-geometry-library-java', version: '0.0.1'
     dependenciesFunctionsJar group: 'com.google.cloud', name: 'google-cloud-storage', version: '1.67.0'
-
+    dependenciesFunctionsJar group: 'org.apache.flink', name: 'flink-shaded-hadoop2-uber', version: '2.4.1-1.8.3'
```
Comment (MayurGubrele) [inline dagger-functions/build.gradle:67]: This jar exists in /opt/flink/lib but with a different version (flink-shaded-hadoop-2-uber-2.8.3-7.0.jar)

File: dagger-functions/src/test/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    public void shouldGetFileNames() {
+        ClassLoader classLoader = getClass().getClassLoader();
+        File file = new File(Objects.requireNonNull(classLoader.getResource("test_udf.py")).getFile());
+        String pathFile = file.getAbsolutePath();
+
+        PythonFileType pythonFileType = new PythonFileType(pathFile);
+        List<String> fileNames = pythonFileType.getFileNames();
+
+        Assert.assertEquals("[test_udf.py]", fileNames.toString());
+    }
```
Comment (prakharmathur82) [inline dagger-functions/src/test/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileTypeTest.java:22]: can we add failure scenario tests, like what happens if pathFile is empty, null?

**Review by prakharmathur82 — CHANGES_REQUESTED** [review]

File: dagger-functions/src/main/java/io/odpf/dagger/functions/exceptions/PythonFilesNullException.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package io.odpf.dagger.functions.exceptions;
+
+/**
+ * The type Python files null exception.
+ */
+public class PythonFilesNullException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (jesrypandawa) [reply dagger-functions/src/main/java/io/odpf/dagger/functions/exceptions/PythonFilesNullException.java:6]: this exception is thrown when the python_files input config is empty/ null. this is handled in PythonUdfManager class.

There are already files not found exception when registering python function, the input config is not null but flink cannot find the file.

These two things are different for me. so I change it to PythonFilesNullException

File: dagger-functions/src/main/java/io/odpf/dagger/functions/udfs/python/file/type/ZipFileType.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+     *
+     * @param pythonFile  the python file
+     * @param fileSourceFactory the file source factory
+     */
+    public ZipFileType(String pythonFile, FileSourceFactory fileSourceFactory) {
+        this.pythonFile = pythonFile;
+        this.fileSourceFactory = fileSourceFactory;
+    }
+
+    @SneakyThrows
```
[the remark this answers is not in the document]
Comment (jesrypandawa) [reply dagger-functions/src/main/java/io/odpf/dagger/functions/udfs/python/file/type/ZipFileType.java:42]: the interface of this class didn't contain `throws IOException`. this is because we are using a method that requires this. there are few options for this actually.
1. adding SneakyThrows
2. add throws IOException on the interface, but PythonFileType will also have this which is not necessary
3. add try-catch inside the method. (personally, I don't like using this :p )

which one do you think we should use?

File: dagger-functions/src/main/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileType.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+     *
+     * @param pythonFile the python file
+     */
+    public PythonFileType(String pythonFile) {
+        this.pythonFile = pythonFile;
+    }
+
+    @Override
+    public List<String> getFileNames() {
+        String name = pythonFile.substring(pythonFile.lastIndexOf('/') + 1);
```
[the remark this answers is not in the document]
Comment (jesrypandawa) [reply dagger-functions/src/main/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileType.java:24]: we actually only need the python filename for registering the udf. so as long as the file is py, we just can get the filename without downloading it. this is different if the format is in zip on gcs, to get the python filenames, we need to get that file first and extracted the files to get python filenames inside that zip.

File: dagger-functions/src/main/java/io/odpf/dagger/functions/udfs/python/file/source/local/LocalFileSource.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    /**
+     * Instantiates a new Local file source.
+     *
+     * @param pythonFileSource the python file
+     */
+    public LocalFileSource(String pythonFileSource) {
+        this.pythonFile = pythonFileSource;
+    }
+
+    @SneakyThrows
```
[the remark this answers is not in the document]
Comment (jesrypandawa) [reply dagger-functions/src/main/java/io/odpf/dagger/functions/udfs/python/file/source/local/LocalFileSource.java:25]: same as above

File: dagger-functions/src/test/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileTypeTest.java (hunk 3 above)
Comment (jesrypandawa) [reply dagger-functions/src/test/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileTypeTest.java:22]: btw, i will change all the variable name of pathFile to pythonFile. if its null, it will throw PythonFilesNullException. this is handled in PythonUdfManagerClass

File: dagger-functions/src/test/java/io/odpf/dagger/functions/udfs/python/file/type/ZipFileTypeTest.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        when(fileSourceFactory.getFileSource()).thenReturn(gcsFileSource);
+        when(gcsFileSource.getGcsClient()).thenReturn(gcsClient);
+        when(gcsFileSource.getObjectFile()).thenReturn(zipInBytes);
+
+        ZipFileType zipFileType = new ZipFileType(pathFile, fileSourceFactory);
+
+        List<String> fileNames = zipFileType.getFileNames();
+
+        Assert.assertEquals("[python_udf/scalar/add.py, python_udf/vectorized/substract.py]", fileNames.toString());
+    }
```
[the remark this answers is not in the document]
Comment (jesrypandawa) [reply dagger-functions/src/test/java/io/odpf/dagger/functions/udfs/python/file/type/ZipFileTypeTest.java:61]: all the files in the zip are not .py --> will not cause anything, it will just pass. because its empty and will not register anything on table env

File: dagger-functions/src/main/java/io/odpf/dagger/functions/exceptions/PythonFilesNullException.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (prakharmathur82) [reply dagger-functions/src/main/java/io/odpf/dagger/functions/exceptions/PythonFilesNullException.java:6]: maybe PythonFilesEmptyException?

File: dagger-functions/src/main/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileType.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (prakharmathur82) [reply dagger-functions/src/main/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileType.java:24]: got it

File: dagger-functions/src/test/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileTypeTest.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+        PythonFileType pythonFileType = new PythonFileType(pythonFile);
+        List<String> fileNames = pythonFileType.getFileNames();
+
+        Assert.assertEquals("[]", fileNames.toString());
+    }
+
+    @Test
+    public void shouldThrowNullPointerExceptionIfPythonFilesIsNull() {
+        expectedEx.expect(NullPointerException.class);
```
Comment (prakharmathur82) [inline dagger-functions/src/test/java/io/odpf/dagger/functions/udfs/python/file/type/PythonFileTypeTest.java:38]: your code should not throw NullPointerException

**Review by prakharmathur82 — CHANGES_REQUESTED** [review]

**Review by prakharmathur82 — APPROVED** [review]
